moveit_core/robot_trajectory/test/robot_trajectory_test.cpp

```diff
@@ -0,0 +1,319 @@
+        "<joint name=\"base_joint\"/>"
+        "</group>"
+        "<group name=\"base_with_bad_subgroups\">"
+        "<group name=\"error\"/>"
+        "</group>"
+        "</robot>";
+
+    urdf::ModelInterfaceSharedPtr urdf_model = urdf::parseURDF(MODEL2);
+    srdf::ModelSharedPtr srdf_model(new srdf::Model());
+    srdf_model->initString(*urdf_model, SMODEL2);
+    robot_model.reset(new moveit::core::RobotModel(urdf_model, srdf_model));
+  }
+
+  void TearDown() override
+  {
+  }
+
+protected:
+  moveit::core::RobotModelConstPtr robot_model;
+};
+
+
+std::size_t generateTestTraj(robot_trajectory::RobotTrajectory& traj,
+                             const moveit::core::RobotModelConstPtr& robot_model,
+                             const robot_model::JointModelGroup* joint_model_group)
+{
+  traj.clear();
+
+  std::shared_ptr<robot_state::RobotState> robot_state(new robot_state::RobotState(robot_model));
+  robot_state->update();
+
+  // 3 waypoints with default joints
+  for (std::size_t traj_ix = 0; traj_ix < 3; ++traj_ix)
+  {
+    // robot_state.reset(new robot_state::RobotState(*robot_state));
+    traj.addSuffixWayPoint(robot_state::RobotStatePtr(new robot_state::RobotState(*robot_state)), 1.0);
+  }
+  return traj.size();
+}
+
+TEST_F(OneRobot, testGenerateTrajectory)
+{
+  const robot_model::JointModelGroup* joint_model_group = robot_model->getJointModelGroup("base_from_base_to_e");
+  robot_trajectory::RobotTrajectory traj(robot_model, joint_model_group);
+
+  // The full trajectory should be of length 3
+  const std::size_t expected_full_traj_len = 3;
+
+  // Generate a test trajectory
+  std::size_t full_traj_len = generateTestTraj(traj, robot_model, joint_model_group);
+
+  // Test the generateTestTraj still generates a trajectory of length 3
+  EXPECT_EQ(full_traj_len, expected_full_traj_len);  // full traj should be 3 waypoints long
+}
+
+TEST_F(OneRobot, testGetFirstWayPointAndAddPrefix)
+{
+  const robot_model::JointModelGroup* joint_model_group = robot_model->getJointModelGroup("base_from_base_to_e");
+  robot_trajectory::RobotTrajectory traj(robot_model, joint_model_group);
+  generateTestTraj(traj, robot_model, joint_model_group);
+
+  robot_state::RobotState new_robot_state(robot_model);
+  new_robot_state.update();
+  traj.addPrefixWayPoint(new_robot_state, 1.0);
+  robot_state::RobotState first_robot_state = traj.getFirstWayPoint();
+
+  EXPECT_EQ(first_robot_state, new_robot_state);
+}
+
+TEST_F(OneRobot, testIterators)
+{
+  const robot_model::JointModelGroup* joint_model_group = robot_model->getJointModelGroup("base_from_base_to_e");
+  robot_trajectory::RobotTrajectory traj(robot_model, joint_model_group);
+  generateTestTraj(traj, robot_model, joint_model_group);
+
+  robot_state::RobotState traj_robot_state(robot_model);
+  traj_robot_state.update();
+
+  for (auto pair : traj)
+  {
+    // Compare waypoints
+    EXPECT_EQ(*pair.first, traj_robot_state);
```

Comment:
Could you compare pointers instead of values here, so you don't need to implement an equality operator for class?